File: AdjMatrix.py
'''
        0
       / \
      1   2
     / \   
    3   4   
'''
class Graph:
    def __init__(self,size):
        self.size = size
        self.adjmat = []
        for i in range(size):
            self.adjmat.append([0 for i in range(size)])
    def AddEdge(self,d1,d2):
        if d1 is d2:
            print("Same value so no edje present")
            return
        self.adjmat[d1][d2] = 1
        self.adjmat[d2][d1] = 1
    def BFS(self,start,visited):
        queue = [start]
        visited[start] = True
        while queue:
            v = queue.pop(0) #Queue uses First in First Out
            print(v, end = " ")
            for i in range(self.size):
                if (self.adjmat[v][i] == 1 and (visited[i] == False)):
                    visited[i] = True
                    queue.append(i)
    def DFS(self,start,visited):
        visited[start] = True
        stack = [start]
        while stack:
            v = stack.pop(-1) #Stack uses First In Last Out
            print(v,end = " ")
            for i in range(self.size):
                if (self.adjmat[v][i] == 1 and visited[i] == False):
                    visited[i] = True
                    stack.append(i)
    def Display(self):
        for row in self.adjmat:
            for val in row:
                print("{}".format(val),end = " ")
            print()
    
num_vert = 5
obj = Graph(num_vert)
obj.AddEdge(0,1)
obj.AddEdge(0,2)
obj.AddEdge(1,3)
obj.AddEdge(1,4)
obj.Display()
print("Bredth First Search")
visited = [False]*num_vert
obj.BFS(0,visited)
print("\nDepth Fisrst Search")
visited = [False]*num_vert
obj.DFS(0,visited)

File: AdjMatrix.cpp
/*
    0
   / \
  1   2
 / \
3   4
*/
#include<bits/stdc++.h>
#include<queue>
#include<stack>
using namespace std;
void AddEdge(vector <vector<int>> &adjmat,int u,int v)
{
    adjmat[u][v] = 1;
    adjmat[v][u] = 1;
}
void display(vector<vector<int>> &adjmat,int num_vert)
{
    for(int i = 0;i<num_vert;i++)
    {
        for(auto x:adjmat[i])
        {
            cout<<x<<" ";
        }
    cout<<endl;
    }
}
void BFS(vector<vector<int>> &adjmat,int start)
{
    vector<int>visited = {start};
    queue<int>q;
    q.push(start);
    while(q.empty() == false)
    {
        int v = q.front();
        cout<<v<<" ";
        q.pop();
        for(int i = 0;i<adjmat.size();i++)
        {
            if(adjmat[v][i] == 1 && count(visited.begin(),visited.end(),i) == false)
            {
                visited.push_back(i);
                q.push(i);
            }
        }
    }
    cout<<endl;
}
void DFS(vector<vector<int>> &adjmat,int start)
{
    vector<int>visited = {start};
    stack<int>s;
    s.push(start);
    while(s.empty() == false)
    {
        int v = s.top();
        cout<<v<<" ";
        s.pop();
        for(int i = 0;i<adjmat.size();i++)
        {
            if(adjmat[v][i] == 1 && count(visited.begin(),visited.end(),i) == false)
            {
                visited.push_back(i);
                s.push(i);
            }
        }
    }
    cout<<endl;
}
int main()
{
    int num_vert = 5;
    vector<vector<int>>adjmat(num_vert,(vector<int>(num_vert,0)));
    AddEdge(adjmat,0,1);
    AddEdge(adjmat,0,2);
    AddEdge(adjmat,1,3);
    AddEdge(adjmat,1,4);
    display(adjmat,num_vert);
    cout<<"Bredth First Search"<<endl;
    BFS(adjmat,0);
    cout<<"Depth First Search"<<endl;
    DFS(adjmat,0);
}